File: packages/shared/src/data/guild.rs
use super::{
    ChannelId, GuildId, UserId,
    channel::prelude::{Channel, ChannelCategory},
    user::UserPublic,
};
use crate::data::{RoleId, permissions::Permissions};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct GuildIdentity {
    pub display_name: String,
    pub avatar_url: Option<String>,
    pub bio: Option<String>,
    pub show_global_username: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Role {
    pub id: RoleId,
    pub name: String,
    pub color: Option<i32>,
    pub hoist: bool,
    pub position: i32,
    pub permissions: i64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct GuildMember {
    pub guild_id: GuildId,
    pub user_id: UserId,
    pub roles: Vec<RoleId>,
    pub joined_at: String,
    pub data: UserPublic,
    pub identity: Option<GuildIdentity>,
}

impl GuildMember {
    pub fn has_permission(&self, all_roles: &[Role], required: Permissions) -> bool {
        let member_permissions: Vec<i64> = all_roles
            .iter()
            .filter(|r| self.roles.contains(&r.id))
            .map(|r| r.permissions)
            .collect();

        let total_perms = Permissions::compute_base_permissions(&member_permissions);
        total_perms.contains(required)
    }

    pub fn is_admin(&self, all_roles: &[Role]) -> bool {
        self.has_permission(all_roles, Permissions::ADMINISTRATOR)
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct GuildProfile {
    pub owner_id: UserId,
    pub banner_url: Option<String>,
    pub icon_url: Option<String>,
    pub name: String,
    pub default_channel_id: Option<ChannelId>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct GuildSummary {
    pub id: GuildId,
    #[serde(flatten)]
    pub profile: GuildProfile,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Guild {
    pub id: GuildId,
    #[serde(flatten)]
    pub profile: GuildProfile,
    pub roles: Vec<Role>,
    pub members: Vec<GuildMember>,
    pub categories: Vec<ChannelCategory>,
    pub channels: Vec<Channel>,
}

File: apps/client/src/types/guild.ts
import type { ChannelId, GuildId, LogId, RoleId, UserId } from ".";
import type { Channel, ChannelCategory } from "./channel";
import type { UserPublic } from "./user";

export type Permissions = number;

export interface GuildIdentity {
  avatar_url: string | null;
  bio: string | null;
  display_name: string;
  show_global_username: boolean;
}

export interface Role {
  color: number | null;
  hoist: boolean;
  id: RoleId;
  name: string;
  permissions: Permissions;
  position: number;
}

export interface GuildMember {
  data: UserPublic;
  guild_id: GuildId;
  identity: GuildIdentity | null;
  joined_at: string;
  roles: RoleId[];
  user_id: UserId;
}

export interface GuildProfile {
  banner_url: string | null;
  default_channel_id: ChannelId | null;
  icon_url: string | null;
  name: string;
  owner_id: UserId;
}

export type GuildSummary = {
  id: GuildId;
} & GuildProfile;

export type Guild = {
  id: GuildId;
  roles: Role[];
  members: GuildMember[];
  categories: ChannelCategory[];
  channels: Channel[];
} & GuildProfile;

export enum AuditActionType {
  GuildUpdate = 1,
  ChannelCreate = 10,
  ChannelUpdate = 11,
  ChannelDelete = 12,
  CategoryCreate = 13,
  CategoryUpdate = 14,
  CategoryDelete = 15,
  MemberKick = 20,
  MemberBanAdd = 22,
  MemberBanRemove = 23,
  MemberUpdate = 24,
  RoleCreate = 30,
  RoleUpdate = 31,
  RoleDelete = 32,
  MessageDelete = 72,
}

export interface AuditLogEntry {
  action_type: AuditActionType;
  actor_id: UserId;
  changes: Record<string, unknown> | null;
  created_at: string;
  guild_id: GuildId;
  id: LogId;
  reason: string | null;
  target_id: string | null;
}
